client/context.go

```diff
@@ -212,27 +212,30 @@ func (ctx Context) PrintString(str string) error {
 // either text or json. If text, toPrint will be YAML encoded. Otherwise, toPrint
 // will be JSON encoded using ctx.JSONMarshaler. An error is returned upon failure.
 func (ctx Context) PrintOutput(toPrint proto.Message) error {
-	return ctx.printOutput(toPrint)
+	// always serialize JSON initially because proto json can't be directly YAML encoded
+	out, err := ctx.JSONMarshaler.MarshalJSON(toPrint)
```

Comment:
One thing that this PR made clear is that:
- ctx.JSONMarshaler == ProtoCodec (afaict, in all instances in the codebase)
- ctx.LegacyAmino == AminoCodec

In this case, would it make sense to rename `ctx.JSONMarshaler` to `ctx.ProtoCodec`, and let it take a concrete type `ProtoCodec`?

I guess my question is: is there a use case where we would want JSONMarshaler to be AminoCodec?

Reply:
`ProtoCodec` does not only implement `JSONMarshaler` but also `BinaryMarshaler`. It seems like only `JSONMarshaler` is needed for the client context.

In the case of amino tests (ie using https://github.com/cosmos/cosmos-sdk/blob/marie/6982-json-marshaler-proto/simapp/params/amino.go#L15), client context `JSONMarshaler` might be `AminoCodec`.

Reply:
`AminoCodec` also implements `JSONMarshaler`. It has all the same methods as `ProtoCodec` and requires `proto.Message`s as input but does amino marshaling instead of proto. So we use `JSONMarshaler` anywhere we want proto _or_ amino.

`LegacyAmino` _does not_ require `proto.Message` so we use it for legacy cases where we _don't have_ `proto.Message`.

Does that clarify @amaurymartiny ?

Reply:
Got it, that makes it clear. Thanks.